File: executioncontext/projectview/src/main/java/org/jetbrains/bsp/bazel/projectview/parser/sections/ProjectViewSingletonSectionParser.java
package org.jetbrains.bsp.bazel.projectview.parser.sections;

import io.vavr.control.Option;
import io.vavr.control.Try;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;
import org.jetbrains.bsp.bazel.projectview.model.sections.ProjectViewSingletonSection;
import org.jetbrains.bsp.bazel.projectview.parser.splitter.ProjectViewRawSections;

/**
 * Implementation of single value section parser.
 *
 * <p>It takes a raw section and trims the content.
 *
 * @param <T> type of parsed single value section
 */
abstract class ProjectViewSingletonSectionParser<V, T extends ProjectViewSingletonSection<V>>
    extends ProjectViewSectionParser<T> {

  private static final Logger log = LogManager.getLogger(ProjectViewSingletonSectionParser.class);

  protected ProjectViewSingletonSectionParser(String sectionName) {
    super(sectionName);
  }

  @Override
  public Option<T> parseOrDefault(ProjectViewRawSections rawSections, Option<T> defaultValue) {
    var section = parse(rawSections);

    logParseOrDefault(section, defaultValue);

    return section.orElse(defaultValue);
  }

  private void logParseOrDefault(Option<T> section, Option<T> defaultValue) {
    if (section.isDefined()) {
      log.debug("Parsed '{}' section. Result:\n{}", sectionName, section);
    } else {
      log.debug("Returning default for '{}' section. Result:\n{}", sectionName, defaultValue);
    }
  }

  @Override
  public Option<T> parse(ProjectViewRawSections rawSections) {
    var section =
        rawSections.getLastSectionWithName(sectionName).map(this::parse).flatMap(Try::get);

    logParse(section);

    return section;
  }

  private void logParse(Option<T> section) {
    log.debug("Parsed '{}' section. Result:\n{}", sectionName, section);
  }

  @Override
  protected Option<T> parse(String sectionBody) {
    return Option.of(sectionBody.strip())
        .filter(body -> !body.isEmpty())
        .map(this::mapRawValue)
        .map(this::createInstance);
  }

  protected abstract V mapRawValue(String rawValue);

  protected abstract T createInstance(V value);
}

File: executioncontext/projectview/src/main/java/org/jetbrains/bsp/bazel/projectview/parser/sections/ProjectViewSingletonSectionParser.kt
package org.jetbrains.bsp.bazel.projectview.parser.sections

import org.apache.logging.log4j.LogManager
import org.jetbrains.bsp.bazel.projectview.model.sections.ProjectViewBazelPathSection
import org.jetbrains.bsp.bazel.projectview.model.sections.ProjectViewSingletonSection
import org.jetbrains.bsp.bazel.projectview.model.sections.ProjectViewBuildManualTargetsSection
import org.jetbrains.bsp.bazel.projectview.model.sections.ProjectViewDeriveTargetsFromDirectoriesSection
import org.jetbrains.bsp.bazel.projectview.model.sections.ProjectViewImportDepthSection
import org.jetbrains.bsp.bazel.projectview.parser.splitter.ProjectViewRawSections
import java.nio.file.Path
import kotlin.io.path.Path

/**
 * Implementation of single value section parser.
 *
 *
 * It takes a raw section and trims the content.
 *
 * @param <T> type of parsed single value section
</T> */
abstract class ProjectViewSingletonSectionParser<V, T : ProjectViewSingletonSection<V>> protected constructor(
    override val sectionName: String
) : ProjectViewSectionParser<T>() {

    override fun parse(rawSections: ProjectViewRawSections): T? =
        rawSections.getLastSectionWithName(sectionName)
            ?.let { parse(it) }
            ?.get()
            .also { log.debug("Parsed '$sectionName' section. Result:\n$it") }

    override fun parse(sectionBody: String): T? =
        sectionBody.trim()
            .ifBlank { null }
            ?.let { mapRawValue(it) }
            ?.let { createInstance(it) }

    protected abstract fun mapRawValue(rawValue: String): V

    protected abstract fun createInstance(value: V): T

    companion object {
        private val log = LogManager.getLogger(ProjectViewSingletonSectionParser::class.java)
    }
}


object ProjectViewBazelPathSectionParser :
    ProjectViewSingletonSectionParser<Path, ProjectViewBazelPathSection>(ProjectViewBazelPathSection.SECTION_NAME) {

    override fun mapRawValue(rawValue: String): Path = Path(rawValue)

    override fun createInstance(value: Path): ProjectViewBazelPathSection = ProjectViewBazelPathSection(value)
}

object ProjectViewDeriveTargetsFromDirectoriesSectionParser :
        ProjectViewSingletonSectionParser<Boolean, ProjectViewDeriveTargetsFromDirectoriesSection>(ProjectViewDeriveTargetsFromDirectoriesSection.SECTION_NAME) {

    override fun mapRawValue(rawValue: String): Boolean = rawValue.toBoolean()

    override fun createInstance(value: Boolean): ProjectViewDeriveTargetsFromDirectoriesSection =
            ProjectViewDeriveTargetsFromDirectoriesSection(value)
}

object ProjectViewBuildManualTargetsSectionParser :
    ProjectViewSingletonSectionParser<Boolean, ProjectViewBuildManualTargetsSection>(ProjectViewBuildManualTargetsSection.SECTION_NAME)
{

    override fun mapRawValue(rawValue: String): Boolean = rawValue.toBoolean()

    override fun createInstance(value: Boolean): ProjectViewBuildManualTargetsSection = ProjectViewBuildManualTargetsSection(value)
}


object ProjectViewImportDepthSectionParser :
    ProjectViewSingletonSectionParser<Int, ProjectViewImportDepthSection>(ProjectViewImportDepthSection.SECTION_NAME) {

    override fun mapRawValue(rawValue: String): Int = rawValue.toInt()

    override fun createInstance(value: Int): ProjectViewImportDepthSection =
        ProjectViewImportDepthSection(value)
}
